File: agent/vehicle_agent/capture.py
import json
from collections.abc import Iterable, Iterator
from datetime import UTC, datetime
from pathlib import Path
from typing import Any, TextIO

from agent.vehicle_agent.models import CANFrame
from agent.vehicle_agent.providers.obdlink import parse_can_frame

CAPTURE_FORMAT = "vehinode-can-jsonl"
CAPTURE_VERSION = 1


class CaptureError(ValueError):
    pass


class CANRecorder:
    def __init__(self, output: TextIO, metadata: dict[str, object]):
        self.output = output
        self.output.write(
            json.dumps(
                {
                    "type": "header",
                    "format": CAPTURE_FORMAT,
                    "version": CAPTURE_VERSION,
                    "created_at": datetime.now(UTC).isoformat(),
                    "metadata": metadata,
                }
            )
            + "\n"
        )

    def write(self, frame: CANFrame) -> None:
        self.output.write(json.dumps(frame.as_dict()) + "\n")
        self.output.flush()


def replay_capture(path: str | Path) -> Iterator[tuple[dict[str, Any], CANFrame]]:
    with Path(path).open() as source:
        first = source.readline()
        try:
            header = json.loads(first)
        except json.JSONDecodeError as exc:
            raise CaptureError("capture header is invalid JSON") from exc
        if header.get("format") != CAPTURE_FORMAT or header.get("version") != CAPTURE_VERSION:
            raise CaptureError("unsupported capture format or version")
        metadata = header.get("metadata", {})
        for number, line in enumerate(source, 2):
            try:
                row = json.loads(line)
                if row.get("type") != "frame":
                    continue
                frame = parse_can_frame(
                    f"{row['can_id']} {row['data']}", timestamp=float(row["timestamp"])
                )
            except (KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
                raise CaptureError(f"invalid frame at line {number}") from exc
            yield metadata, frame


def record_frames(output: TextIO, frames: Iterable[CANFrame], metadata: dict[str, object]) -> int:
    recorder = CANRecorder(output, metadata)
    count = 0
    for frame in frames:
        recorder.write(frame)
        count += 1
    return count

File: agent/internal/capture/capture.go
package capture

import (
	"bufio"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/Buco7854/vehinode/agent/internal/model"
)

const Format = "vehinode-can-jsonl"
const Version = 1

type Recorder struct{ output io.Writer }

func NewRecorder(output io.Writer, metadata map[string]any) (*Recorder, error) {
	recorder := &Recorder{output: output}
	header := map[string]any{"type": "header", "format": Format, "version": Version, "created_at": time.Now().UTC(), "metadata": metadata}
	return recorder, recorder.write(header)
}

func (recorder *Recorder) Write(frame model.CANFrame) error {
	return recorder.write(map[string]any{"type": "frame", "timestamp": frame.Timestamp, "can_id": fmt.Sprintf("0x%03X", frame.CANID), "data": strings.ToUpper(hex.EncodeToString(frame.Data))})
}

func (recorder *Recorder) write(value any) error {
	content, err := json.Marshal(value)
	if err != nil {
		return err
	}
	content = append(content, '\n')
	_, err = recorder.output.Write(content)
	return err
}

type Capture struct {
	Metadata map[string]any
	Frames   []model.CANFrame
}

func Read(path string) (Capture, error) {
	file, err := os.Open(path)
	if err != nil {
		return Capture{}, err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 4096), 1024*1024)
	if !scanner.Scan() {
		return Capture{}, fmt.Errorf("capture header is missing")
	}
	var header struct {
		Format   string         `json:"format"`
		Version  int            `json:"version"`
		Metadata map[string]any `json:"metadata"`
	}
	if err := json.Unmarshal(scanner.Bytes(), &header); err != nil || header.Format != Format || header.Version != Version {
		return Capture{}, fmt.Errorf("unsupported capture format or version")
	}
	result := Capture{Metadata: header.Metadata}
	line := 1
	for scanner.Scan() {
		line++
		var row struct {
			Type      string  `json:"type"`
			Timestamp float64 `json:"timestamp"`
			CANID     string  `json:"can_id"`
			Data      string  `json:"data"`
		}
		if err := json.Unmarshal(scanner.Bytes(), &row); err != nil {
			return Capture{}, fmt.Errorf("invalid frame at line %d: %w", line, err)
		}
		if row.Type != "frame" {
			continue
		}
		id, err := strconv.ParseInt(strings.TrimPrefix(row.CANID, "0x"), 16, 32)
		if err != nil {
			return Capture{}, fmt.Errorf("invalid frame at line %d", line)
		}
		data, err := hex.DecodeString(row.Data)
		if err != nil || len(data) > 8 {
			return Capture{}, fmt.Errorf("invalid frame at line %d", line)
		}
		result.Frames = append(result.Frames, model.CANFrame{Timestamp: row.Timestamp, CANID: int(id), Data: data})
	}
	return result, scanner.Err()
}
